### ISSUE TITLE
setting lr, lr criterion, lr_decay, lr_decay_steps and left/right context

### ISSUE THREAD

**Issue by omprakashsonie** [facebookresearch/wav2letter#813]
### Question
I am using transformer architecture with CTC.

1. Wanted to try out different learning rate, lr_decay, lr_decay_steps, step_size, gamma, lrcosine which have following default values:

    -lr_decay (Epoch for the first LR decay) type: int64
      default: 9223372036854775807
    -lr_decay_step (Epochs for each new LR decay) type: int64
      default: 9223372036854775807
    -stepsize (We multiply LR by gamma every stepsize updates) type: int64
      default: 9223372036854775807
    -gamma (the LR annealing multiplier) type: double default: 1

    -lrcosine (use cosine learning rate schedule) type: bool default: false

Can you pls provide some explanation on using above flags?

2.  Is there a way to set left/right context for transformer? or is it not applicable in case of transformer as it will take full context?
    -rightWindowSize (right median window width) type: int64 default: 50
    -leftWindowSize (left median window width) type: int64 default: 50

#### Additional Context
    -lr (learning rate) - using 0.4
    -lrcrit (criterion learning rate) - using 0.4

**Comment by tlikhomanenko** [comment]
> Can you pls provide some explanation on using above flags?

lr_decay is the epoch number when we start do divide current learning rate by too, we repeat this division every lr_decay_step epochs after: for example for lr_decay=100, lr_decay_step=50 and lr=1, from 1-99 lr will be 1, at 100-149 lr will be 0.5, 150-200 lr=0.025, ...

About cosine and gamma: both are about scaling lr in addition to the lr_decay. Either cosine or gamma policies are used. In case of cosine is set to false, we use gamma and scaler of lr will be gamma ^ (current_update / stepsize). In case of cosine scaler is cos(current_batch / nbatches * pi / 2).

Here is the code on the lr policies https://github.com/facebookresearch/wav2letter/blob/v0.2/Train.cpp#L579.

> Is there a way to set left/right context for transformer? or is it not applicable in case of transformer as it will take full context?

Currently we don't support this. Which part of context do you mean (there could be different options depending on the ctc or s2s models)? Right now transformer module and transformer in s2s decoder part are using full context.

**Comment by omprakashsonie** [comment]
Thank you very much for your help and explanation.

**Comment by tlikhomanenko** [comment]
> Sorry for bothering you.
> I read the [paper about MLS](https://arxiv.org/pdf/2012.03411.pdf), and notice that `decay learning rate by a factor of 2 each time the WER reaches a plateau on the validation sets`. I'm just curious where I can find the `decaying LR mechanism` described in this paper. Or maybe there is more detailed README file that I ignored.
> Looking forward to your replay.
> Thanks.

This phrase exactly describe the decaying mechanism. As soon as we see no decreasing in WER on validation set we set `new_lr = old_lr / 2` and continue training while we again don't see WER decreasing. We often repeat this step 2 times.

**Comment by xiaosdawn** [comment]
So, I understand that this machanism is done manually. Am I right?

**Comment by tlikhomanenko** [comment]
Either manually but if you knows the general dynamic of training (or if you are ok to decay) you can use these flags https://github.com/flashlight/flashlight/blob/master/flashlight/app/asr/common/Flags.cpp#L157-L164

**Comment by xiaosdawn** [comment]
Thanks. I know the flags. But if I don't know the specific epoch that the WER reaches a plateau on the validtion sets, I will not set the specific lr_decay.

**Comment by tlikhomanenko** [comment]
> Thanks. I know the flags. But if I don't know the specific epoch that the WER reaches a plateau on the validtion sets, I will not set the specific lr_decay.

Yep. You can use also cosine schedule or add any other schedules for lr. But in most other cases you need to know other hyper-params, like total number of updates, etc. So anyway either you use someone's recipe or run with several hyper-params to know what is the best. In my practice at least for speech recognition decaying lr after a plateau was a good strategy. You can even not decay but do exponential moving averaging.

**Comment by xiaosdawn** [comment]
I got it. Thanks very much.
